### PR TITLE
Add tests

### PR DESCRIPTION
**Pull request by mrakitin**
`pytest` tests that are based on the existing examples.
The PR branch is off the `reorg` branch.

Closes https://github.com/Synchrotron-Radiation-Workshop/SRW/issues/8.

### CODE REVIEW

File: .github/workflows/testing.yml [hunk 1]
```diff
@@ -64,6 +64,14 @@ jobs:
...
+          cd python
+          pytest -s -vv --import-mode=importlib srwpy/tests/ -k fast
+
+      - name: Run tests for slow examples
+        run: |
+          set -vxeuo pipefail
+          conda install -y -c conda-forge mpi4py openmpi
+          cd examples/python
+          cat SRWLIB_Example08.py | sed 's/nMacroElec = 50000/nMacroElec = 100/' > SRWLIB_Example08_100_particles.py
+          mpirun -n 4 python SRWLIB_Example08_100_particles.py
```
Comment (mrakitin) [inline .github/workflows/testing.yml]:
```suggestion
          mpirun --oversubscribe -n 4 python SRWLIB_Example08_100_particles.py
```

File: .github/workflows/testing.yml [hunk 2]
```diff
@@ -64,6 +64,14 @@ jobs:
...
-          # cp -rv python/srwpy/tests tests
-          # pytest -s -vv --import-mode=importlib python/srwpy/tests/
-          pytest -s -vv --pyargs srwpy
+          cd python
+          pytest -s -vv --import-mode=importlib srwpy/tests/ -k fast
+
+      - name: Run tests for slow examples
+        run: |
+          set -vxeuo pipefail
+          conda install -y -c conda-forge mpi4py openmpi
```
Comment (mrakitin) [inline .github/workflows/testing.yml]:
```suggestion
          conda install -y -c conda-forge mpi4py mpi
```

File: .github/workflows/testing.yml [hunk 3]
```diff
@@ -64,6 +64,14 @@ jobs:
...
+          pytest -s -vv --import-mode=importlib srwpy/tests/ -k fast
+
+      - name: Run tests for slow examples
+        run: |
+          set -vxeuo pipefail
+          conda install -y -c conda-forge mpi4py mpi
+          cd examples/python
+          cat SRWLIB_Example08.py | sed 's/nMacroElec = 50000/nMacroElec = 100/' > SRWLIB_Example08_100_particles.py
+          mpirun --oversubscribe -n 4 python SRWLIB_Example08_100_particles.py
+          cat __srw_logs__/srwl_stat_wfr_emit_prop_multi_e_*.log
```
Comment (mrakitin) [inline .github/workflows/testing.yml]:
```suggestion
          cat __srwl_logs__/srwl_stat_wfr_emit_prop_multi_e_*.log
```

File: python/srwpy/tests/test_example12.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import os
```
Comment (mrakitin) [inline python/srwpy/tests/test_example12.py]:
```suggestion
import os
import sys
```

File: python/srwpy/tests/test_example12.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+import sys
+import subprocess
+import pytest
+
+
+@pytest.mark.fast
+@pytest.mark.skipif(sys.platform == "linux", reason="runs only on Linux")
```
Comment (mrakitin) [inline python/srwpy/tests/test_example12.py]:
```suggestion
@pytest.mark.skipif(sys.platform != "linux", reason="runs only on Linux")
```

**Comment by mrakitin** [comment]
@hhslepicka, are you around to give a brief review and merge?

**Comment by mrakitin** [comment]
@himanshugoel2797, @ahebnl, do you mind having a look to spot any issues? Thanks!

**Comment by hhslepicka** [comment]
@mrakitin I will take a look... I am out tomorrow but if you could wait until Monday I can do it by next week.

**Comment by ahebnl** [comment]
Hi Maksim,
This workflow looks very cool. I can't wait to try it and learn 'pytest'
through this new SRW version. I will try it these days and let you know if
I have any questions.
Thanks and best,
An

**Review by hhslepicka — APPROVED** [review]
The changes here look good to me. I sent some questions over Slack. Other than that, all good.

File: python/tests/conftest.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    print(f"Output file: {output_file}")
+
+    lines = raw_example_lines
+
+    if hasattr(request, "param"):
+        param_name = "nMacroElec"
+        param_old_value = 50000
+        param_new_value = request.param
+
+
```
Comment (mrakitin) [inline python/tests/conftest.py]:
```suggestion
```

File: python/tests/test_example01.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
+# Imports from the example:
+from __future__ import print_function #Python 2.7 compatibility
```
Comment (mrakitin) [inline python/tests/test_example01.py]:
```suggestion
from __future__ import print_function  # Python 2.7 compatibility
```

File: python/tests/test_example02.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
+# Imports from the example:
+from __future__ import print_function #Python 2.7 compatibility
```
Comment (mrakitin) [inline python/tests/test_example02.py]:
```suggestion
from __future__ import print_function  # Python 2.7 compatibility
```

File: python/tests/test_example03.py [hunk 9]
```diff
@@ -0,0 +1,25 @@
+# Imports from the example:
+from __future__ import print_function #Python 2.7 compatibility
```
Comment (mrakitin) [inline python/tests/test_example03.py]:
```suggestion
from __future__ import print_function  # Python 2.7 compatibility
```

File: python/tests/test_example04.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
+# Imports from the example:
+from __future__ import print_function #Python 2.7 compatibility
```
Comment (mrakitin) [inline python/tests/test_example04.py]:
```suggestion
from __future__ import print_function  # Python 2.7 compatibility
```

File: python/tests/test_example05.py [hunk 11]
```diff
@@ -0,0 +1,25 @@
+# Imports from the example:
+from __future__ import print_function #Python 2.7 compatibility
```
Comment (mrakitin) [inline python/tests/test_example05.py]:
```suggestion
from __future__ import print_function  # Python 2.7 compatibility
```

File: python/tests/test_example06.py [hunk 12]
```diff
@@ -0,0 +1,41 @@
+# Imports from the example:
+from __future__ import print_function #Python 2.7 compatibility
```
Comment (mrakitin) [inline python/tests/test_example06.py]:
```suggestion
from __future__ import print_function  # Python 2.7 compatibility
```

File: python/tests/test_example07.py [hunk 13]
```diff
@@ -0,0 +1,26 @@
+# Imports from the example:
+from __future__ import print_function #Python 2.7 compatibility
```
Comment (mrakitin) [inline python/tests/test_example07.py]:
```suggestion
from __future__ import print_function  # Python 2.7 compatibility
```

File: python/tests/test_example07.py [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+from srwpy.srwlib import *
+from srwpy.uti_plot import * #required for plotting
```
Comment (mrakitin) [inline python/tests/test_example07.py]:
```suggestion
from srwpy.uti_plot import *  # required for plotting
```

File: python/tests/test_example09.py [hunk 15]
```diff
@@ -0,0 +1,26 @@
+# Imports from the example:
+from __future__ import print_function #Python 2.7 compatibility
```
Comment (mrakitin) [inline python/tests/test_example09.py]:
```suggestion
from __future__ import print_function  # Python 2.7 compatibility
```

File: python/tests/test_example09.py [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+from srwpy.srwlib import *
+from srwpy.uti_plot import * #required for plotting
```
Comment (mrakitin) [inline python/tests/test_example09.py]:
```suggestion
from srwpy.uti_plot import *  # required for plotting
```

File: python/tests/test_example13.py [hunk 17]
```diff
@@ -0,0 +1,24 @@
+# Imports from the example:
+from __future__ import print_function #Python 2.7 compatibility
```
Comment (mrakitin) [inline python/tests/test_example13.py]:
```suggestion
from __future__ import print_function  # Python 2.7 compatibility
```

File: python/tests/test_example13.py [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+from srwpy.srwlib import *
+from srwpy.uti_plot import * #required for plotting
```
Comment (mrakitin) [inline python/tests/test_example13.py]:
```suggestion
from srwpy.uti_plot import *  # required for plotting
```

File: python/tests/test_example14.py [hunk 19]
```diff
@@ -0,0 +1,26 @@
+# Imports from the example:
+from __future__ import print_function #Python 2.7 compatibility
```
Comment (mrakitin) [inline python/tests/test_example14.py]:
```suggestion
from __future__ import print_function  # Python 2.7 compatibility
```

File: python/tests/test_example14.py [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+from srwpy.srwlib import *
+from srwpy.uti_math import *
+from srwpy.uti_plot import * #required for plotting
```
Comment (mrakitin) [inline python/tests/test_example14.py]:
```suggestion
from srwpy.uti_plot import *  # required for plotting
```

File: python/tests/test_example17.py [hunk 21]
```diff
@@ -0,0 +1,32 @@
+# Imports from the example:
+from __future__ import print_function #Python 2.7 compatibility
```
Comment (mrakitin) [inline python/tests/test_example17.py]:
```suggestion
from __future__ import print_function  # Python 2.7 compatibility
```

File: python/tests/test_example17.py [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+from srwpy.srwlib import *
+from srwpy.srwl_uti_smp import *
+from srwpy.uti_plot import * #required for plotting
```
Comment (mrakitin) [inline python/tests/test_example17.py]:
```suggestion
from srwpy.uti_plot import *  # required for plotting
```

File: python/tests/test_example18.py [hunk 23]
```diff
@@ -0,0 +1,29 @@
+# Imports from the example:
+from __future__ import print_function #Python 2.7 compatibility
```
Comment (mrakitin) [inline python/tests/test_example18.py]:
```suggestion
from __future__ import print_function  # Python 2.7 compatibility
```

File: python/tests/test_example18.py [hunk 24]
```diff
@@ -0,0 +1,29 @@
...
+from srwpy.srwlib import *
+from srwpy.srwl_uti_smp import *
+from srwpy import srwl_uti_smp_rnd_obj3d
+from srwpy.uti_plot import * #required for plotting
```
Comment (mrakitin) [inline python/tests/test_example18.py]:
```suggestion
from srwpy.uti_plot import *  # required for plotting
```

File: python/tests/test_example06.py [hunk 25]
```diff
@@ -0,0 +1,41 @@
...
+
+    # end of example, start testing
+    assert len(arIS) == 20000
+    assert len(arI) == 40401
+    for obj in [stkF, stkP]:
+        assert hasattr(obj, "mesh")
+        for param in ["zStart",
+                    "eStart", "eFin", "ne",
+                    "xStart", "xFin", "nx",
+                    "yStart", "yFin", "ny"]:
```
Comment (mrakitin) [inline python/tests/test_example06.py]:
```suggestion
                      "eStart", "eFin", "ne",
                      "xStart", "xFin", "nx",
                      "yStart", "yFin", "ny"]:
```

File: python/tests/test_example17.py [hunk 26]
```diff
@@ -0,0 +1,32 @@
...
+
+    import PIL
+    assert type(imgProc) is PIL.Image.Image
+    assert imgProc.size == (1079, 700)
+
+    assert hasattr(wfr, "mesh")
+    for param in ["zStart",
+                "eStart", "eFin", "ne",
+                "xStart", "xFin", "nx",
+                "yStart", "yFin", "ny"]:
```
Comment (mrakitin) [inline python/tests/test_example17.py]:
```suggestion
                  "eStart", "eFin", "ne",
                  "xStart", "xFin", "nx",
                  "yStart", "yFin", "ny"]:
```

File: python/tests/test_example18.py [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+    # end of example, start testing
+    assert len(arI0) == 6084
+    assert len(arI1) == 4194304
+    assert len(arLogI1) == 4194304
+
+    assert hasattr(wfr, "mesh")
+    for param in ["zStart",
+                "eStart", "eFin", "ne",
+                "xStart", "xFin", "nx",
+                "yStart", "yFin", "ny"]:
```
Comment (mrakitin) [inline python/tests/test_example18.py]:
```suggestion
                  "eStart", "eFin", "ne",
                  "xStart", "xFin", "nx",
                  "yStart", "yFin", "ny"]:
```

File: python/tests/test_example19.py [hunk 28]
```diff
@@ -0,0 +1,29 @@
+# Imports from the example:
+from __future__ import print_function #Python 2.7 compatibility
```
Comment (mrakitin) [inline python/tests/test_example19.py]:
```suggestion
from __future__ import print_function  # Python 2.7 compatibility
```

File: python/tests/test_example19.py [hunk 29]
```diff
@@ -0,0 +1,29 @@
...
+from srwpy.srwlib import *
+from srwpy.srwl_uti_smp import *
+from srwpy import srwl_uti_smp_rnd_obj3d
+from srwpy.uti_plot import * #required for plotting
```
Comment (mrakitin) [inline python/tests/test_example19.py]:
```suggestion
from srwpy.uti_plot import *  # required for plotting
```

File: python/tests/test_example19.py [hunk 30]
```diff
@@ -0,0 +1,29 @@
...
+    # end of example, start testing
+    assert len(arI0) == 6084
+    assert len(arI1) == 4194304
+    assert len(arLogI1) == 4194304
+
+    assert hasattr(wfr, "mesh")
+    for param in ["zStart",
+                "eStart", "eFin", "ne",
+                "xStart", "xFin", "nx",
+                "yStart", "yFin", "ny"]:
```
Comment (mrakitin) [inline python/tests/test_example19.py]:
```suggestion
                  "eStart", "eFin", "ne",
                  "xStart", "xFin", "nx",
                  "yStart", "yFin", "ny"]:
```

**Review by mrakitin — COMMENTED** [review]
Final touch-ups.

**Comment by mrakitin** [comment]
@ahebnl, @himanshugoel2797, @ochubar, ready for your review. I plan to merge it this week and then address #7 to restructure the files/dirs as we discussed.

@hhslepicka, if you have a moment, could you also check the PR? I simplified the tests (with better fixtures, so fewer repetitions are made), and also wrapped the slow tests into `pytest`.

**Comment by hhslepicka** [comment]
@mrakitin it looks great to me. The new simplified tests are nice! Feel free to merge it into `reorg`.

**Comment by mrakitin** [comment]
Thanks, @hhslepicka!

@ahebnl informed me that she tried to install the package from this branch in a new conda env, and the tests ran successfully. An, if you would like to go through the code changes, and let me know if you spot any issues with this (or have questions), I'd appreciate it. Thank you.

**Comment by ahebnl** [comment]
Sure. I will let you know if I have any questions.
Thank, An

**Comment by mrakitin** [comment]
I am going to merge it today by COB to proceed with #7. Let me know if there are any concerns regarding the updates in this PR.

**Pull request merged**